File: daily/cpp/977_squares_of_sorted_array.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  vector<int> sortedSquares(vector<int>& nums) {
    /// Two pointer approach O(N)
    int n = nums.size();
    int idx = n - 1;
    int le = 0, ri = n - 1;
    vector<int> res(n);
    while (le <= ri) {
      if (abs(nums[le]) > abs(nums[ri])) {
        res[idx] = nums[le] * nums[le];
        ++le;
      } else if (abs(nums[le]) <= abs(nums[ri])) {
        res[idx] = nums[ri] * nums[ri];
        --ri;
      }
      --idx;
    }
    return res;

    /// Using sort O(NlogN)
  }
};

File: daily/rust/977_squares_of_sorted_array.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; nums.len()];
        let (mut le, mut ri, mut idx) = (0, nums.len() - 1, nums.len() - 1);
        while le <= ri {
            if nums[le].abs() > nums[ri].abs() {
                res[idx] = nums[le] * nums[le];
                le += 1;
            } else if nums[le].abs() <= nums[ri].abs() {
                res[idx] = nums[ri] * nums[ri];
                ri -= 1;
            }
            idx -= 1;
        }
        res
    }
}